code/src/pages/cart/CartPage.test.tsx

```diff
@@ -0,0 +1,103 @@
+import { screen } from "@testing-library/react";
+
+import CartPage from "@/pages/cart/CartPage";
+import { useGetCartItemsQuery } from "@/store/api/cartApi";
+import { useUserDetailsSelector } from "@/store/slices/userSlice";
+import renderWithProviders from "@/utils/render-with-providers/renderWithProviders";
+
+jest.mock("@/store/slices/userSlice", () => ({
+  useUserDetailsSelector: jest.fn()
+}));
+
+jest.mock("@/store/api/cartApi", () => ({
+  useGetCartItemsQuery: jest.fn()
+}));
+
+jest.mock("@/utils/format-price/formatPrice");
+
+const mockUseUserDetailsSelector = useUserDetailsSelector as jest.Mock;
+const mockUseGetCartItemsQuery = useGetCartItemsQuery as jest.Mock;
+
+const mockedCartItems = {
+  items: [
+    {
+      productId: "8efbee82-8a0c-407a-a4c0-16bbad40a23e",
+      image: "https://example.com/phone.jpg",
+      name: "Iphone",
+      productPrice: 100.45,
+      quantity: 2,
+      calculatedPrice: 200.9
+    }
+  ],
+  totalPrice: 1000.45
+};
+
+describe("CartPage", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    mockUseUserDetailsSelector.mockReturnValue({ id: "4455667788" });
+  });
+
+  test("renders loading state", () => {
+    mockUseGetCartItemsQuery.mockReturnValue({
+      data: null,
+      isLoading: true,
+      error: false
+    });
+
+    renderWithProviders(<CartPage />);

```

Comment:
Make helper function to reuse it, because if is repeated

Reply:
> Make helper function to reuse it, because if is repeated

I tried but can't. Show me your suggestion for it please and I'll rewrite it

Reply:
```jsx

const renderWithMock = ({data = null, isLoading = false, error = false}) => {
    mockUseGetCartItemsQuery.mockReturnValue({
      data,
      isLoading,
      error
    });

    renderWithProviders(<CartPage />);
}
```

Reply:
> ```js
> const renderWithMock = ({data = null, isLoading = false, error = false}) => {
>     mockUseGetCartItemsQuery.mockReturnValue({
>       data,
>       isLoading,
>       error
>     });
> 
>     renderWithProviders(<CartPage />);
> }
> ```

It just function for my case. And why we not created it before.
Create please reusable function. I'm not see the reason used it in one place

Reply:
You use similar logic in 3 tests. This should be refactored

Reply:
> You use similar logic in 3 tests. This should be refactored. Check other tests, we always try to reuse logic in tests

Just for your 2 tests it will be ok ? Huh. 3 it's not 2 tests)
Reconsider your review approach and try to follow your own suggestions
FYI if you have similar `renderWithProviders` just pass it into `beforeEach` block and if it's not depend with mock different responses data from hook etc.

I won't to create utility function because that there is no sense in this..

<img width="954" alt="image" src="https://github.com/user-attachments/assets/5d997c5a-9478-4ed5-9e1e-fc65dc6ce135">

Reply:
You provided an example that confirms your point. I will propose to consider another example. Take a look on `DashboardTabs.test.tsx` for example:

```jsx
const renderWithMockSearchParams = (params?: Record<string, string>) => {
  (useSearchParams as jest.Mock).mockReturnValue([
    new URLSearchParams(params),
    mockSetSearchParams
  ]);

  renderWithProviders(<DashboardTabs />);
};
```

Same with `ProductsPage`, `OrdersPage` written by @D1nviz, `ProductsContainer` and `BestSellers` by @GeorgeShvab, also `useSignIn`, `useSignUp` for example. So, that confirms, that in all places where there was reusable code that differs very little, a setup or render local function was added. Ideally yes, my example could be refactored. But the important note is that one line with `renderWithProviders` in example from my code that you provided does not make a big deal for readability concerns but in your case you just copied and pasted 7 lines of code over and over again (as much as 3 times). You contradict yourself, because you claim that you should avoid code repetition.

In my opinion it should be refactored and I even provided and example on how you should do it.

P.S. I will add rule `you should avoid code duplication as much as possible in your code and your tests` and `try to keep our tests clear and create setup or render local functions to improve code readability` into our readme if it is not obvious for you

Reply:
> You provided an example that confirms your point. I will propose to consider another example. Take a look on `DashboardTabs.test.tsx` for example:
> 
> ```js
> const renderWithMockSearchParams = (params?: Record<string, string>) => {
>   (useSearchParams as jest.Mock).mockReturnValue([
>     new URLSearchParams(params),
>     mockSetSearchParams
>   ]);
> 
>   renderWithProviders(<DashboardTabs />);
> };
> ```
> 
> Same with `ProductsPage`, `OrdersPage` written by @D1nviz, `ProductsContainer` and `BestSellers` by @GeorgeShvab, also `useSignIn`, `useSignUp` for example. So, that confirms, that in all places where there was reusable code that differs very little, a setup or render local function was added. Ideally yes, my example could be refactored. But the important note is that one line with `renderWithProviders` in example from my code that you provided does not make a big deal for readability concerns but in your case you just copied and pasted 7 lines of code over and over again (as much as 3 times). You contradict yourself, because you claim that you should avoid code repetition.
> 
> In my opinion it should be refactored and I even provided and example on how you should do it.
> 
> P.S. I will add rule `you should avoid code duplication as much as possible in your code and your tests` and `try to keep our tests clear and create setup or render local functions to improve code readability` into our readme if it is not obvious for you

Changed. But please also follow this rules from your side as I talked about `beforeEach` block in your tests case

Reply:
Looks much better, thanks! What about mine test? Should I create separate issue for it now or we can fix it in connection to some PR?

Reply:
> Looks much better, thanks! What about mine test? Should I create separate issue for it now or we can fix it in connection to some PR?

When you'll work with another issue. Don't forget to change. That's not urgent

Reply:
I created new issue and self-assigned to it in order to definitely avoid forgetting